lib/Doctrine/DBAL/Platforms/MySqlPlatform.php

```diff
@@ -385,7 +385,7 @@ public function getListTableColumnsSQL($table, $database = null)
         }
 
         return "SELECT COLUMN_NAME AS Field, COLUMN_TYPE AS Type, IS_NULLABLE AS `Null`, ".
-               "COLUMN_KEY AS `Key`, COLUMN_DEFAULT AS `Default`, EXTRA AS Extra, COLUMN_COMMENT AS Comment, " .
+               "COLUMN_KEY AS `Key`, TRIM('\'' FROM QUOTE(COLUMN_DEFAULT)) AS `Default`, EXTRA AS Extra, COLUMN_COMMENT AS Comment, " .
```

Comment:
Is the problem really in the introspection query? If I add the following assertion to the newly introduced test,
```php
$this->_conn->executeUpdate('INSERT INTO string_escaped_default_value VALUES()');
$fetched = $this->_conn->executeQuery('SELECT def_string from string_escaped_default_value')
    ->fetchColumn();
$this->assertSame($value, $fetched);
```
it fails as:
```
1) Doctrine\Tests\DBAL\Functional\Schema\MySqlSchemaManagerTest::testEscapedDefaultValueMustBePreserved
Failed asserting that two strings are identical.
--- Expected
+++ Actual
@@ @@
-a\0a\'a"a
+aa'a"a
 a
-a	a\Za\\a
+a	aa\a
```
Which probably means that the actual default value is not what the tests defines and then retrieves.

Reply:
I was under the impression that `getListTableColumnsSQL()` was supposed to return SQL that could be used directly, without any further escaping. And more specifically that they could be compared directly to the value coming from user-defined mapping via `Comparator::diffTable()`. Since that user-defined value must be escaped, the introspected value should also be escaped. That's why I chose to do the escaping in there. 

But maybe we can see the issue from a different point of view and always diff un-escaped values ? so that means that the user-defined value coming from the mapping should be un-escaped ? But since escaping is platform specific where/how should we un-escaped user-defined values ?

Reply:
> I was under the impression that `getListTableColumnsSQL()` was supposed to return SQL that could be used directly, without any further escaping.

Designed this way, it would allow stored SQL injections, so no.

> But maybe we can see the issue from a different point of view and always diff un-escaped values?

Yes, the values should be only escaped when building SQL containing them.

> So that means that the user-defined value coming from the mapping should be un-escaped? 

Yes, because escaping is platform-specific.

> But since escaping is platform specific where/how should we un-escaped user-defined values?

Why are user-defined values escaped?

Reply:
> Why are user-defined values escaped?

Because if I want my default value to be `Foo\Bar\Baz`, I will have to escape the `\` and thus writing something like `@ORM\Column(options={"default" = "Foo\\Bar\\Baz"}))`.

I see that for a similar problem with quoted table names, that it was resolved in `AbstractAsset::_setName()` where quotes are detected and removed before comparison.

Should we adopt a similar strategy and modify `Column::setDefault()` to automatically un-escape strings ? If so should we somehow detect which strings need to be un-escaped ? or un-escape all of them and then introducing a breaking change ?

Another approach would be to write the mapping non-escaped, and let Doctrine do the escaping automatically in all cases. But that would directly contradict what is done for table names and would also introduce a breaking change. So I don't think it's the best way to solve that.

The more I think about this, the more I think we won't be able to fix it without introducing a potential breaking change. But maybe there is yet another way... ? Maybe the original way to escape introspected value is less breaking after all ?

Reply:
> Because if I want my default value to be `Foo\Bar\Baz`, I will have to escape the `\` and thus writing something like `@ORM\Column(options={"default" = "Foo\\Bar\\Baz"}))`.

It's the escaping required by the PHP string syntax, the value of the default does not contain escaped slashes when used in defining the schema:
```php
echo "Foo\\Bar\\Baz", PHP_EOL;

// Foo\Bar\Baz
```

> I see that for a similar problem with quoted table names, that it was resolved in `AbstractAsset::_setName()` where quotes are detected and removed before comparison.
> 
> Should we adopt a similar strategy and modify `Column::setDefault()` to automatically un-escape strings? If so should we somehow detect which strings need to be un-escaped ? or un-escape all of them and then introducing a breaking change?

I don't know the implementation details of the above, but there shouldn't be any un-escaping in any layer. The un-escaping is only supposed to be done by the parser of the language in whose syntax the escaping is done. Like in my code example, the un-escaping is done by PHP.

> Another approach would be to write the mapping non-escaped, and let Doctrine do the escaping automatically in all cases. But that would directly contradict what is done for table names and would also introduce a breaking change. So I don't think it's the best way to solve that.

This is the only right way how things should work. None values should be escaped anywhere else then before concatenation with SQL/HTML/whatever "program". Furthermore, in the SQL context, if a value is bound to a prepared statement, it must not be escaped because it's not parsed.